h/security/permits/engine.py

```diff
@@ -0,0 +1,69 @@
+from enum import Enum
+
+
+def predicate(requires=None):
+    def decorator(function):
+        function.requires = requires or []
+        return function
+
+    return decorator
+
+
+class PredicateBasedPermissions:
+    def __init__(self, mapping):
+        self.mapping = self._expand_clauses(mapping)
+
+    def permits(self, identity, context, permission):
+        if clauses := self.mapping.get(permission):
+            statement_cache = {}
+
+            for clause in clauses:
+                if self._clause_true(clause, statement_cache, identity, context):
+                    return True
+
+            return False
+
+        return False
+
+    def _clause_true(self, clause, statement_cache, identity, context):
+        for predicate in clause:
+            value = statement_cache.get(predicate)
+            if value is None:
+                if isinstance(predicate, Enum):
+                    # We are going to assume this predicate is a permission
+                    value = self.permits(identity, context, predicate)
+                else:
+                    value = predicate(identity, context)
+
+                statement_cache[predicate] = value
+
+            if not value:
+                return False
+
+        return True
+
```

Comment:
Everything below here is dealing with the predicate requires stuff. The actual code of the evaluation is above.